### PR TITLE
Add support for InputNumber and additional constraints. Fixes #5165

### LINKED ISSUE #5165: InputNumber and Spinner: support additional bean validation constraints
## 1) Environment
- PrimeFaces version: 7.0
- Does it work on the newest sources in GitHub? (Build by source -> https://github.com/primefaces/primefaces/wiki/Building-From-Source) No
- Application server + version: all
- Affected browsers: all

## 2) Issue Description
Right now, only Spinner is considered in [BeanValidationInputMetadataTransformer](https://github.com/primefaces/primefaces/blob/70a71f9524dd68d065198c9c442d80253274c9c1/src/main/java/org/primefaces/metadata/transformer/impl/BeanValidationInputMetadataTransformer.java) :
https://github.com/primefaces/primefaces/blob/70a71f9524dd68d065198c9c442d80253274c9c1/src/main/java/org/primefaces/metadata/transformer/impl/BeanValidationInputMetadataTransformer.java#L91-L102
InputNumber should be included too.
Adding `@DecimalMax` and `@DecimalMin` (and maybe `@Digits`) annotations support would be useful too.

### PR DESCRIPTION
**Pull request by NicolaIsotta**
Fixes #5165

### CODE REVIEW

File: src/main/java/org/primefaces/metadata/transformer/impl/BeanValidationInputMetadataTransformer.java [hunk 1]
```diff
@@ -91,23 +99,64 @@ protected void applyConstraint(ConstraintDescriptor constraintDescriptor, UIInpu
...
                 Min min = (Min) constraint;
                 spinner.setMin(min.value());
             }
+            if (annotationType.equals(DecimalMax.class) && spinner.getMax() == Double.MAX_VALUE) {
+                DecimalMax max = (DecimalMax) constraint;
+                try {
+                    spinner.setMax(Double.parseDouble(max.value()));
+                }
+                catch (NumberFormatException ex) {
+                }
```
Comment (melloware) [inline src/main/java/org/primefaces/metadata/transformer/impl/BeanValidationInputMetadataTransformer.java:117]: I would add `LOGGER.log(Level.WARN, "Failed setting Spinner max value: " + ex.getMessage());` I don't like empty try catches.

File: src/main/java/org/primefaces/metadata/transformer/impl/BeanValidationInputMetadataTransformer.java [hunk 2]
```diff
@@ -91,23 +99,64 @@ protected void applyConstraint(ConstraintDescriptor constraintDescriptor, UIInpu
...
+            }
+            if (annotationType.equals(DecimalMin.class) && spinner.getMin() == Double.MIN_VALUE) {
+                DecimalMin min = (DecimalMin) constraint;
+                try {
+                    spinner.setMin(Double.parseDouble(min.value()));
+                }
+                catch (NumberFormatException ex) {
+                }
```
Comment (melloware) [inline src/main/java/org/primefaces/metadata/transformer/impl/BeanValidationInputMetadataTransformer.java:126]: I would add `LOGGER.log(Level.WARN, "Failed setting Spinner min value: " + ex.getMessage());` I don't like empty try catches.

**Review by melloware — CHANGES_REQUESTED** [review]

**Review by tandraschko — COMMENTED** [review]
Is DecimalMin in BVal 1.0?

**Comment by melloware** [comment]
> Is DecimalMin in BVal 1.0?

Yes DecimalMin and Max are in 1.0.

**Comment by tandraschko** [comment]
+1
But please also change the Date to LocalDate in this class

**Comment by NicolaIsotta** [comment]
> +1
> But please also change the Date to LocalDate in this class

Sure. Two questions:
- Since time could be in the calendar pattern, maybe LocalDateTime is better?
- Should I consider BV 2.0 new time constraints (`@PastOrPresent` and `@FutureOrPresent`)? We can use class names like in OutputLabelRenderer so no need to change dependency.

**Comment by tandraschko** [comment]
1) Not sure. CC @christophs78
2) yes. We can not upgrade as PF would not work on JavaEE6/7 then

**Comment by christophs78** [comment]
Calendar and DatePicker currently (7.1-dev; Master-Branch) supports LocalDate as minDate and maxDate-values. (And java.util.Date for backward-compatibility-reasons.)
See [https://primefaces.github.io/primefaces/7_1/#/components/datepicker?id=datepicker](https://primefaces.github.io/primefaces/7_1/#/components/datepicker?id=datepicker)
Maybe we should add support for LocalDateTime for minDate and maxDate-Attributes for convenience-reasons. And for edge-cases where Calendar and DatePicker are used with showTime=true. And maybe we should support LocalTime for timeOnly=true.

**Comment by NicolaIsotta** [comment]
Updated with LocalDate e BV 2.0 new time contraints.

**Comment by tandraschko** [comment]
+1 
Can someone retest it?
I assume it also works with DatePicker?

**Comment by melloware** [comment]
It should work with DatePicker, @NicolaIsotta can you confirm and then I will take a pass at testing it.

**Comment by NicolaIsotta** [comment]
It's not ready to be tested as DatePicker and LocalDateTime/java.util.Date combination seems a bit broken ATM (see #4625 #4626 and relative fixes)

**Comment by tandraschko** [comment]
Whats happening?
@christophs78 can you assist pls?

**Comment by christophs78** [comment]
@NicolaIsotta: Applied a fix/work-around for 4625 (for LocalDateTime) during my work on java.time-support. Please look into [https://github.com/primefaces/showcase-facelift/blob/master/src/main/webapp/ui/input/datePickerJava8.xhtml](https://github.com/primefaces/showcase-facelift/blob/master/src/main/webapp/ui/input/datePickerJava8.xhtml). Compare this with [https://github.com/primefaces/showcase-facelift/blob/master/src/main/webapp/ui/input/datePicker.xhtml](https://github.com/primefaces/showcase-facelift/blob/master/src/main/webapp/ui/input/datePicker.xhtml)
Could you look please look into this and give me feedback? (I know this solution is not ideal, but it´s a work-around without needing to to a deepdive into the clientside JS.) 

@tandraschko, @mertsincan: Is it possible to deploy the 7.1-dev-Showcase? Think it would be ideal to have the 7.0-Release Showcase deployed and in parallel the 7.1-dev-Showcase?

@NicolaIsotta: I know there is some room for improvement (as already mentioned) for Type-Support for minDate/maxDate-Attributes. I will work on this sometime in october.

**Comment by NicolaIsotta** [comment]
I'll open a new issue for datepicker to avoid cluttering this conversation.

> @NicolaIsotta: I know there is some room for improvement (as already mentioned) for Type-Support for minDate/maxDate-Attributes. I will work on this sometime in october.

mindate and maxdate already support LocalDateTime, no need for additional work.

**Comment by tandraschko** [comment]
+1 
Can you merge plz melloware?

**Pull request merged**